transformations/transformations/block_index_transformations.py

```diff
@@ -0,0 +1,428 @@
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+# In applying this licence, ECMWF does not waive the privileges and immunities
+# granted to it by virtue of its status as an intergovernmental organisation
+# nor does it submit to any jurisdiction.
+
+from loki import (
+    Transformation, ProcedureItem, ir, Module, as_tuple, SymbolAttributes, BasicType, Variable,
+    RangeIndex, Array, FindVariables, resolve_associates, SubstituteExpressions, FindNodes,
+    recursive_expression_map_update
+)
+
+from transformations.single_column_coalesced import SCCBaseTransformation
+
+__all__ = ['BlockViewToFieldViewTransformation', 'BlockIndexInjectTransformation']
+
+class BlockViewToFieldViewTransformation(Transformation):
+    """
+    A very IFS-specific transformation to replace per-block, i.e. per OpenMP-thread, view pointers with per-field
+    view pointers. It should be noted that this transformation only replaces the view pointers but does not actually
+    insert the block index into the promoted view pointers. Therefore this transformation must always be followed by
+    the :any:`BlockIndexInjectTransformation`.
+
+    For example, the following code:
+
+    .. code-block:: fortran
+
+        do jlon=1,nproma
+          mystruct%p(jlon,:) = 0.
+        enddo
+
+    is transformed to:
+
+    .. code-block:: fortran
+
+        do jlon=1,nproma
+          mystruct%p_field(jlon,:) = 0.
+        enddo
+
+    Where the rank of ``my_struct%p_field`` is one greater than that of ``my_struct%p``. Specific arrays in individual
+    routines can also be marked for exclusion from this transformation by assigning them to the `exclude_arrays` list
+    in the :any:`SchedulerConfig`.
+
+    Parameters
+    ----------
+    horizontal : :any:`Dimension`
+        :any:`Dimension` object describing the variable conventions used in code
+        to define the horizontal data dimension and iteration space.
+    global_gfl_ptr: bool
+        Toggle whether thread-local gfl_ptr should be replaced with global.
+    key : str, optional
+        Specify a different identifier under which trafo_data is stored
+    """
+
+    _key = 'BlockViewToFieldViewTransformation'
+    """Default identifier for trafo_data entry"""
+
+    item_filter = (ProcedureItem,)
+
+    def __init__(self, horizontal, global_gfl_ptr=False, key=None):
+        self.horizontal = horizontal
+        self.global_gfl_ptr = global_gfl_ptr
+        if key:
+            self._key = key
+
+    @staticmethod
+    def get_parent_typedef(var, symbol_map):
+        """Utility method to retrieve derived-tyoe definition of parent type."""
+
+        if not var.parent.type.dtype.typedef is BasicType.DEFERRED:
+            return var.parent.type.dtype.typedef
+        if  (_parent_type := symbol_map.get(var.parent.type.dtype.name, None)):
+            if not _parent_type.type.dtype.typedef is BasicType.DEFERRED:
+                return _parent_type.type.dtype.typedef
```

Comment:
Let me explain with a minimal example:

```
use field_array_module, only : field_3rb_array
...
type(field_3rb_array) :: yda_array
...
yda_array%p = 0.
```

Once we create the "dummy" field_api definition and enrich the subroutine with it, the `use` statement will now point to the correct `TypeDef`, but `var.parent.type.dtype.typedef` will still be `BasicType.DEFERRED` for the lhs of the assignment.

To update those definitions we would have to do a recursive `._update` on all the nodes in the body, which would be seriously overkill. So I put that workaround in to retrieve the correct type information from the symbol map.

Reply:
The `BlockViewToFieldView` transformations updates the type information in the routine body wherever appropriate, so that subsequent transformations in the pipeline can retrieve the correct `TypeDef` simply with `var.parent.type.dtype.typedef`.

Reply:
OK, that's roughly what I understood as well.
But since you're only after the updated type info in the symbol table, you don't need to rebuild the expression nodes as long as the symbol table is correctly updated. For the successors, you're already calling `ir.enrich` and you could do a corresponding `routine.enrich(definitions)` in `process_driver`. 

If that's not enough to update the symbol table entries, then that sounds like a deficiency of the `enrich` method to me.

Reply:
The problem arises from the fact that we don't store `DataType`'s in the symbol table. Referring again to the above example, `field_3rb_array` is indeed stored in the symbol table, but before enrichment this is a `DeferredTypeSymbol` and afterwards it is a `ProcedureSymbol`. The useful information is in `field_3rb_array.type.dtype`, which we don't store directly as a symbol.

`yda_array.type.dtype` therefore (correctly) does not point to the imported symbol called `field_3rb_array`. So even when that import is enriched, the `DataType` information in the variable declarations and expression nodes is not updated.

For now I've amended `ProgramUnit.enrich` to also update the symbols in variable declaration with the correct `type.dtype`. We probably want to add a mechanism to add `DataType` symbols directly in the symbol table so that simply enriching the imports should be enough.